### PR TITLE
Development

### PR DESCRIPTION
**Pull request by esennesh**

### CODE REVIEW

**Comment by zmheiko** [comment]
Hey @esennesh , I'm wondering on what the rationale is for storing (a pointer to) the distribution object and the distribution arguments in the random variable? Both can be easily recovered for the instantiated distribution object.

**Comment by esennesh** [comment]
> Hey @esennesh , I'm wondering what the rationale is for storing (a pointer to) the distribution object and the distribution arguments in the random variable? Both can be easily recovered for the instantiated distribution object.

Not if you want to automatically rebuild the distribution object without knowing which parameterization was originally given!  I use this to implement resampling over traces.

**Comment by zmheiko** [comment]
> > Hey @esennesh , I'm wondering what the rationale is for storing (a pointer to) the distribution object and the distribution arguments in the random variable? Both can be easily recovered for the instantiated distribution object.
> 
> Not if you want to automatically rebuild the distribution object without knowing which parameterization was originally given! I use this to implement resampling over traces.

Hey Eli, thanks for elaborating. I think its actually still possible to retrieve the parameters even without knowing what they were in the first place:
   ` param_dict = {k: rv.dist.__dict__[k].detach() for k, _ in rv.dist.arg_constraints.items() if k in dist.__dict__}`

Also, can you please explain why this is necessary for resampling? My understanding is that, for each sample value that you could possibly resample you already know the corresponding density under the RV's associated pdf. Thus, when you resample you can just generate the ancestor indices by a method of your choice and then collect the corresponding values from `RV.value` and `RV.log_prob`. There should be no need to actually reevaluate the density of a sample. Am I missing something?
